tests/regression.rs

```diff
@@ -0,0 +1,380 @@
+//!
+//! For remaining types see: ./regression_opcodes.rs
+//!
+//! If you find a type defined in `rust-bitcoin` that implements `Serialize` and does _not_ have a
+//! regression test please add it.
+//!
+//! Types/tests were found using, and are ordered by, the output of: `git grep -l Serialize`.
+//!
+
+#![allow(unused_imports)] // All imports are unused when use-serde feature is not set.
+
+extern crate bitcoin;
+extern crate bincode;
+extern crate secp256k1;
+
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::io::{self, Read, BufReader};
+use std::io::BufRead;
+use std::str::FromStr;
+
+use bincode::serialize;
+
+use bitcoin::{Address, Network, Script, Block, TxIn, TxOut, Transaction, PrivateKey, PublicKey, SchnorrSig, SchnorrSigHashType, EcdsaSigHashType, EcdsaSig, Txid, OutPoint};
+use bitcoin::blockdata::witness::Witness;
+use bitcoin::util::bip32::{ExtendedPrivKey, ExtendedPubKey, ChildNumber, KeySource};
+use bitcoin::util::uint::{Uint128, Uint256};
+use bitcoin::util::taproot::{ControlBlock, TaprootBuilder, LeafVersion};
+use bitcoin::util::psbt::{Input, Output, PartiallySignedTransaction};
+use bitcoin::util::psbt::raw::{self, Pair, Key, ProprietaryKey};
+use bitcoin::util::ecdsa;
+use bitcoin::hashes::hex::FromHex;
+use bitcoin::hashes::{sha256d, Hash, ripemd160, hash160, sha256};
+use bitcoin::consensus::encode::deserialize;
+
+/// Reads a comma separated list of integers from file `name` into a vector. Integers represent the
+/// expected bincode for a serialized data type.
+#[cfg(feature = "serde")]
+fn bincode_from_file(name: &str) -> Vec<u8> {
+    let name = format!("tests/data/bincode/{}", name);
+    let file = File::open(name).unwrap();
+    let reader = BufReader::new(file);
+
+    reader.lines().next().unwrap().unwrap()
+        .split(',').map(|s| s.trim())
+        .filter(|s| !s.is_empty())
+        .map(|s| s.parse().unwrap())
+        .collect()
+}
+
+/// Parses `hex` string into a vector of bytes.
+#[cfg(feature = "serde")]
+fn parse_hex(hex: &str) -> Vec<u8> {
+    Vec::from_hex(hex.trim()).unwrap()
+}
+
+/// Implicitly does regression test for `BlockHeader` also.
+#[test]
+#[cfg(feature = "serde")]
+fn serde_regression_block() {
+    let segwit = include_bytes!("data/testnet_block_000000000000045e0b1660b6445b5e5c5ab63c9a4f956be7e1e69be04fa4497b.raw").to_vec();
+    let block: Block = deserialize(&segwit).unwrap();
+    let got = serialize(&block).unwrap();
+    let want = bincode_from_file("block");
+    assert_eq!(got, want)
+}
+
+#[test]
+#[cfg(feature = "serde")]
+fn serde_regression_script() {
+    let script = Script::from(vec![0u8, 1u8, 2u8]);
+    let got = serialize(&script).unwrap();
+    let want = bincode_from_file("script");
+    assert_eq!(got, want)
+}
+
+#[test]
+#[cfg(feature = "serde")]
+fn serde_regression_txin() {
+    let s = include_str!("data/hex/txin");
+    let txin: TxIn = deserialize(&parse_hex(s)).unwrap();
```

Comment:
Why not just `include_bytes!()` and feed them in directly without decoding?

Reply:
Cool, so now I have binary data in the `tests/data/` directory and am using `include_bytes!`.  Makes the code a bit neater but I fear it makes adding new tests harder because I had to convert all the bincode integer arrays to binary. I've left in a helper function for doing so for other devs (or future Tobin).

Reply:
How about adding a simple shell script as an external too? Or if it's too involved for a shell script, [cargo xtask](https://github.com/matklad/cargo-xtask)

Reply:
I think I've addressed this by moving the utility function to `util.rs`. With the docs I added its pretty easy to add new tests now.